src/main/drivers/dshot_command.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#ifdef USE_DSHOT
+
+#include "common/time.h"
+
+#include "drivers/io.h"
+#include "drivers/motor.h"
+#include "drivers/time.h"
+#include "drivers/timer.h"
+
+#include "drivers/dshot.h"
+#include "drivers/dshot_dpwm.h"
+#include "drivers/dshot_command.h"
+#include "drivers/pwm_output.h"
+
+#define DSHOT_INITIAL_DELAY_US 10000
+#define DSHOT_COMMAND_DELAY_US 1000
+#define DSHOT_ESCINFO_DELAY_US 12000
+#define DSHOT_BEEP_DELAY_US 100000
+#define DSHOT_MAX_COMMANDS 3
+
+typedef enum {
+    DSHOT_COMMAND_STATE_IDLEWAIT,   // waiting for motors to go idle
+    DSHOT_COMMAND_STATE_STARTDELAY, // initial delay period before a sequence of commands
+    DSHOT_COMMAND_STATE_ACTIVE,     // actively sending the command (with optional repeated output)
+    DSHOT_COMMAND_STATE_POSTDELAY   // delay period after the command has been sent
+} dshotCommandState_e;
+
+typedef struct dshotCommandControl_s {
+    dshotCommandState_e state;
+    uint32_t nextCommandCycleDelay;
+    timeUs_t delayAfterCommandUs;
+    uint8_t repeats;
+    uint8_t command[MAX_SUPPORTED_MOTORS];
+} dshotCommandControl_t;
+
+static timeUs_t dshotCommandPidLoopTimeUs = 125; // default to 8KHz (125us) loop to prevent possible div/0
+                                                 // gets set to the actual value when the PID loop is initialized
+
+static dshotCommandControl_t commandQueue[DSHOT_MAX_COMMANDS + 1];
+static uint8_t commandQueueHead;
+static uint8_t commandQueueTail;
+
+void dshotSetPidLoopTime(uint32_t pidLoopTime)
+{
+    dshotCommandPidLoopTimeUs = pidLoopTime;
+}
+
+static FAST_CODE bool dshotCommandQueueFull()
+{   
+    return (commandQueueHead + 1) % (DSHOT_MAX_COMMANDS + 1) == commandQueueTail;
```

Comment:
`mod` and `div` by non-power-of-two are optimized poorly, it may better to write explicit cases for `commandQueueHead > commandQueueTail` and `<=` depending on usage scenario.
This function is marked for ITCM, so in my opinion it makes sense to optimize it further.

Reply:
@joelucid Can you work this out?

Reply:
@DieHertz, ``DSHOT_MAX_COMMANDS`` is defined as ``3`` so the ``mod`` operation is by a power of two.